### PR TITLE
Update contributing docs to be more comprehensive (and fix links)

### PR DESCRIPTION
**Pull request by mfisher87**
This PR depends on #249 to avoid pre-commit.ci autofixes from polluting the review. **Do not merge until #249 has been merged.**

This PR:

* Documents virtual environment creation with pip and micromamba
* Adds troubleshooting section documenting conflict between Nx and a common dotfile management technique
* Fixes links that are currently broken on the live contributing docs: https://geojupyter.github.io/jupytergis/contributing.html

### CODE REVIEW

**Comment by martinRenou** [comment]
It does not go further than that? 

Can you try manually running `yarn install && yarn run build` instead of through that script? See it that changes things or if the logs are more digest

**Comment by mfisher87** [comment]
>It does not go further than that?

Correct, got stuck for about 30 minutes.

> Can you try manually running `yarn install && yarn run build` instead of through that script? See it that changes things or if the logs are more digest

```$ yarn --version
3.8.7

$ yarn install
➤ YN0000: ┌ Resolution step
➤ YN0061: │ rimraf@npm:3.0.2 is deprecated: Rimraf versions prior to v4 are no longer supported
➤ YN0061: │ abab@npm:2.0.6 is deprecated: Use your platform's native atob() and btoa() methods instead
➤ YN0061: │ eslint@npm:8.57.1 is deprecated: This version is no longer supported. Please see https://eslint.org/version-support for other options.
➤ YN0061: │ glob@npm:7.2.3 is deprecated: Glob versions prior to v9 are no longer supported
➤ YN0061: │ json-schema-ref-parser@npm:9.0.9 is deprecated: Please switch to @apidevtools/json-schema-ref-parser
➤ YN0061: │ glob@npm:7.1.7 is deprecated: Glob versions prior to v9 are no longer supported
➤ YN0061: │ inflight@npm:1.0.6 is deprecated: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
➤ YN0061: │ @humanwhocodes/config-array@npm:0.13.0 is deprecated: Use @eslint/config-array instead
➤ YN0061: │ @humanwhocodes/object-schema@npm:2.0.3 is deprecated: Use @eslint/object-schema instead
➤ YN0061: │ popper.js@npm:1.16.1 is deprecated: You can find the new Popper v2 at @popperjs/core, this package is dedicated to the legacy v1
➤ YN0002: │ @deathbeds/jupyterlab-rjsf@npm:1.1.0 doesn't provide codemirror (p4a616), requested by react-codemirror2
➤ YN0002: │ @deathbeds/jupyterlab-rjsf@npm:1.1.0 doesn't provide react (pb33a8), requested by @rjsf/core
➤ YN0002: │ @deathbeds/jupyterlab-rjsf@npm:1.1.0 doesn't provide react (pae0da), requested by react-codemirror2
➤ YN0002: │ @jupytergis/base@workspace:packages/base doesn't provide ol (p7f7a3), requested by ol-pmtiles
➤ YN0002: │ @jupytergis/base@workspace:packages/base doesn't provide react-dom (p87a65), requested by @naisutech/react-tree
➤ YN0002: │ @jupytergis/base@workspace:packages/base doesn't provide react-dom (p01f68), requested by styled-components
➤ YN0002: │ @jupytergis/base@workspace:packages/base doesn't provide react-is (pba653), requested by styled-components
➤ YN0002: │ @jupytergis/jupytergis-root@workspace:. doesn't provide react (p63d4f), requested by @fortawesome/react-fontawesome
➤ YN0002: │ @jupytergis/schema@workspace:packages/schema doesn't provide webpack (pc32ce), requested by source-map-loader
➤ YN0002: │ @jupytergis/schema@workspace:packages/schema doesn't provide webpack (p59ba2), requested by ts-loader
➤ YN0002: │ @jupyterlab/codemirror@npm:3.6.8 doesn't provide yjs (p28a14), requested by y-codemirror
➤ YN0002: │ @jupyterlab/completer@npm:4.3.3 doesn't provide react (pba8ea), requested by @jupyterlab/settingregistry
➤ YN0002: │ @jupyterlab/services@npm:7.3.3 doesn't provide react (pd9c48), requested by @jupyterlab/settingregistry
➤ YN0002: │ babel-plugin-styled-components@npm:2.1.4 [b470c] doesn't provide @babel/core (p53375), requested by @babel/plugin-syntax-jsx
➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
➤ YN0000: └ Completed in 5s 560ms
➤ YN0000: ┌ Fetch step
➤ YN0013: │ yjs-widgets@npm:0.3.8 can't be found in the cache and will be fetched from the remote registry
➤ YN0013: │ yjs@npm:13.6.20 can't be found in the cache and will be fetched from the remote registry
➤ YN0013: │ yocto-queue@npm:0.1.0 can't be found in the cache and will be fetched from the remote registry
➤ YN0013: │ zstddec@npm:0.1.0 can't be found in the cache and will be fetched from the remote registry
➤ YN0013: │ typescript@npm:5.7.2 can't be found in the cache and will be fetched from the remote registry
➤ YN0000: └ Completed in 4s 661ms
➤ YN0000: ┌ Link step
➤ YN0007: │ nx@npm:20.2.2 [78682] must be built because it never has been before or the last one failed
➤ YN0007: │ core-js-pure@npm:3.39.0 must be built because it never has been before or the last one failed
➤ YN0007: │ @fortawesome/fontawesome-free@npm:5.15.4 must be built because it never has been before or the last one failed
➤ YN0007: │ es5-ext@npm:0.10.64 must be built because it never has been before or the last one failed
<stuck>
```

:sob:

**Comment by mfisher87** [comment]
Trying more things. `dev-install.sh` also fails on Python 3.13 attempting to build pydantic-core:

```
Collecting pydantic-core==2.10.1 (from pydantic==2.4.2->-r /home/mfisher/Projects/geojupyter/jupytergis/requirements-build.txt (line 7))
  Downloading pydantic_core-2.10.1.tar.gz (347 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [6 lines of output]
      
      Cargo, the Rust package manager, is not installed or is not on PATH.
      This package requires Rust and Cargo to compile extensions. Install it through
      the system's package manager or via https://rustup.rs/
      
      Checking for Rust toolchain....
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.
```

I thought this might be because there is no wheel available for Pydantic in Python 3.13, but it looks like there is :confused: 

Downgrading to Python 3.12 resolves.

**Comment by mfisher87** [comment]
My next experiment changed a few variables and got the same result. 

0) Using a different computer, much closer to a fresh Ubuntu install.
1) Using conda to manage the environment instead of `uv venv`: `conda create -n jupytergis pip yarn=3 python=3.12`.
2) Following the contributing instructions exactly instead of using `uv pip`.

Perhaps next step is to create a docker image which repros this issue.

**Comment by brichet** [comment]
I can't reproduce it.
Can you try cleaning your directory with the following command ? 
`git clean -fdx`

**warning**: it will remove all files not staged in git. If you have some files that you don't want to remove you should stage them before, or stash them.

**Comment by mfisher87** [comment]
I was able to reproduce on a fresh clone. It looks like the issue is that I'm using node 22 :) #114 

Re-testing with Node 20 and 21, will add this information as an important admonition in the contributing doc.

**Comment by gjmooney** [comment]
I'm pretty sure #114 had to do with a screenshot downloading script we used to have. I just tried an install in a new environment and didn't have any issues with python 3.12 and node 22.11. I used micromamba to install JupyterLab and python 3.12 though.

**Comment by mfisher87** [comment]
Interesting, thanks @gjmooney ! I was unable to get past the linking step just by switching to node 20 and 21. I _was_ able to do so by using a conda virtual environment. Which I had tried before! Not sure what I did differently :) I'll test some more and document what I learn!

Thanks everyone for your help getting unstuck :)

**Comment by mfisher87** [comment]
Reproducing this seems to depend on where in my filesystem my project is. From my jupytergis project directory in my home dir, I get this hang whether I'm using a conda virtual environment or venv, whether I've installed yarn via conda or not, regardless of node version. Similarly, I'm able to successfully run the full `dev-install.py` script from `/tmp/jupytergis` with conda, with venv, with node 22, 21, and 20, and without installing yarn (the jlpm-bundled version, 3.5, is used).

There must be some yarn config file that's being picked up from the directory structure above my jupytergis project directory. Perhaps I'll diff the output of `strace -e open` to compare the runs.

**Comment by mfisher87** [comment]
Confirmed that this issue reproduces on a fresh clone in any subdir of my home directory, and **does not** reproduce in any other filesystem location.

**Comment by mfisher87** [comment]
This Dockerfile reproduces the issue:

```Dockerfile
FROM ubuntu:latest

RUN apt update
RUN apt install -y git python3 python3-venv python-is-python3 nodejs

WORKDIR /home/someuser
RUN git clone https://github.com/geojupyter/jupytergis

WORKDIR /home/someuser/jupytergis

RUN python -m venv /opt/venv
ENV PATH="/opt/venv/bin:$PATH"

# This line reproduces the issue!
# This is a common pattern for managing dotfiles with git. My dotfiles repo explains more:
# https://github.com/mfisher87/dotfiles
RUN echo "*" > /home/someuser/.gitignore

RUN python -m pip install jupyterlab

RUN python scripts/dev-install.py
```

**Comment by mfisher87** [comment]
Okay... here's my minimal reproduction repo: https://github.com/mfisher87/sscce-nx-with-bare-repo-dotfiles-management-hangs-indefinitely

It's Nx interacting with my dotfile management strategy. I found related issues: https://github.com/nrwl/nx/issues/27494#issuecomment-2481207598

File: docs/contributing.rst [hunk 1]
```diff
@@ -52,3 +63,14 @@ Development uninstall
...
+
+Troubleshooting
+---------------
+
+* Setup of development environment hangs indefinitely when running the
+  ``dev-install.py`` step, specifically on the Yarn linking step.
+
+  * This may be caused by having a ``.gitignore`` file in your home directory.
+    This is a `known issue with Nx <https://github.com/nrwl/nx/issues/27494>`_.
+    The `only known workaround <https://github.com/nrwl/nx/issues/27494#issuecomment-2481207598>`_ is to remove the ``.gitignore`` file from your home directory or to work in a location outside of the home directory tree.
```
Comment (martinRenou) [inline docs/contributing.rst:76]: 😓

File: docs/contributing.rst [hunk 2]
```diff
@@ -7,22 +7,32 @@ Contributing
...
 .. code-block:: bash
 
     # Clone the repo to your local environment
     git clone https://github.com/geojupyter/jupytergis.git
+
     # Change directory to the jupytergis directory
     cd jupytergis
+
+    # Create a virtual environment and activate it
+    python -m venv .venv
```
Comment (martinRenou) [inline docs/contributing.rst:25]: Can I suggest also adding instructions using `micromamba`? Using a general purpose package manager makes it way easier to also install non-python packages like `yarn` and `qgis`:

```bash
micromamba create --name jupytergis_dev -c conda-forge yarn=3 "nodejs<22" jupyterlab pip qgis

micromamba activate jupytergis_dev
```

File: docs/contributing.rst (hunk 1 above)
Comment (mfisher87) [reply docs/contributing.rst:76]: Right? :laughing:

File: docs/contributing.rst (hunk 2 above)
Comment (mfisher87) [reply docs/contributing.rst:25]: Maybe yarn is required for the `jlpm run watch` script. `jlpm build` works fine for me w/o yarn installed separately, but `jlpm run watch` fails:

```
—————————————————————————————————————————————————————
   ✖  @jupytergis/jupytergis-lab:build                                                         
      /bin/sh: 1: yarn: not found                              
      /bin/sh: 1: yarn: not found                                                              
                                                                                                                                                                                               
                                                                                               
                                                                                               
—————————————————————————————————————————————————————
   ✖  @jupytergis/jupytergis-qgis:build                                                        
      /bin/sh: 1: yarn: not found               
      /bin/sh: 1: yarn: not found
```

Looks like this is controlled by `lerna.json`.

Comment (mfisher87) [reply docs/contributing.rst]: 874b4cb7ef45aae79ce0dffcd9fe24e56adae9e9

4660151014f89851314a96b5fca28e223b39a8b4

Comment (martinRenou) [reply docs/contributing.rst]: > jlpm build works fine for me w/o yarn installed

Right! We can actually consider this a bug a replace [this](https://github.com/geojupyter/jupytergis/blob/32ae6a338d5ce37d1201e7cd2006629909f6ae76/lerna.json#L2) with `jlpm`, that would be less error prone than depending on the yarn version people have installed.

**Review by martinRenou — APPROVED** [review]
Thanks!

**Pull request merged**